### ISSUE TITLE
Add an LRScheduler interface for torch schedulers.

### ISSUE THREAD

**Issue by OliverFM** [pytorch/pytorch#67760]
## 🚀 Feature
Add an `LRScheduler` interface to all schedulers featuring, at a minimum, an abstract property `optimizer` and an abstract method `step`.

## Motivation

I have three reasons for opening this issue:

1. The type system is not clear -- it would be much easier to do static type checking and work with torch if we can abstract away the details of scheduler implementation when we don't need them.
2. When working on fixing #67696 I found that the testing logic is tricky to implement in a clean an reusable way as one must can't verify that the optimizer you think a scheduler is using is actualyl being used. This feature would improve the quality of the tests that we can write.
2. When using the OneCycle and CyclicLR schedulers at work, I ran into linting and typing issues with mypy (unfortunately I don't have a reproducible example at hand) as we could not specify that we wanted to use anything other than an`_LRScheduler` which is a private class.

## Pitch

Add an interface with some basic methods and properties we expect to have in a scheduler.

e.g.
```
import abc

class LRScheduler(abc.ABC):
    @abc.abstractmethod
    @property
    def optimizer(self):
        pass


    @abc.abstractmethod
    def step(self, *args, **kwargs):
        pass
```

one example where this could not be implemented in `_LRScheduler` is in `ChainedScheduler` -- this could be implemented as follows:
```
@property
def optimizer(self):
    # the validity check that is in __init__ can be moved here instead
    return self._schedulers[0].optimizer
```
[source](https://pytorch.org/docs/stable/_modules/torch/optim/lr_scheduler.html#ChainedScheduler)

Doing this ought to make working with torch easier -- when docs talk about using a scheduler, they have a clear interface to point to. This will make type checking easier too!

## Alternatives

The only alternative I have considered is making a different typing hierarchy.
It is not clear, to me at least, what the scope of the optimizer interface should be, but I think the proposal covers a good minimum.

## Additional context

I would be happy to implement this feature, if it is approved.

**Comment by OliverFM** [comment]
One more point -- it is possible that the lack of the call to `super().__init__` in `ChainedScheduler` is a bug, in which case, fixing that would unblock the testing, as it looks like every scheduler has `self.optimizer` as an instance variable.

**Comment by vadimkantorov** [comment]
Maybe actually a contradicting usecase: https://github.com/pytorch/pytorch/issues/67761, but IMO schedulers need a redesign :(

**Comment by OliverFM** [comment]
> Maybe actually a contradicting usecase: #67761, but IMO schedulers need a redesign :(

@vadimkantorov that is a good point -- I think that, whatever the design, maintaining a good interface would help make this all clearer.

**Comment by GdoongMathew** [comment]
Hi, I was thinking if its ok to lay down some limitations of `LRScheduler` if we want to further refactor this module?
I'll first put some of my thoughts here.

1. One `LRScheduler` can only pair with one `Optimizer`
2. Multiple `LRScheduler` can contribute to the parameter groups in the same `Optimizer`, probably we could use something like `ComposedLRScheduler` as the abstract wrapper class, and let its subclass to implement its `step` method.
3. Change the input signature of `step` method to something more adaptive, like `step(self, **kwargs)`
4. Not sure if we could extend the functionality of `LRScheduler` to not just modify `lr`s in an `Optimizer`, but also some of its other hyperparameters, such as `momentum` or `weight_decay`? But I've no clue how to do this elegently.

To extend the original [proposal](https://github.com/pytorch/pytorch/issues/67760#issue-1043376038)
```python
class SchedulerBase(abc.ABC):

    def __init__(self, optimizer: Optimizer):
        self.optimizer = optimizer

    @abc.abstractmethod
    def step(self, **kwargs):
        raise NotImplementedError

    @property
    def optimizer(self) -> Optimizer:
        return self._optimizer

    @optimizer.setter
    def optimizer(self, optimizer: Optimizer):
        if not isinstance(optimizer, Optimizer):
            raise TypeError(f'{type(optimizer).__name__} is not an Optimizer')
        self._optimizer = optimizer

    @property
    def last_lr(self) -> Sequence[float]:
        return self._last_lr

    @last_lr.setter
    def last_lr(self, value: Sequence[float]):
        if not isinstance(value, Sequence):
            raise TypeError(f'Expected a sequence of floats, but got {type(value).__name__}')
        self._last_lr = value


class ComposedScheduler(SchedulerBase):
    """ChainedScheduler and SequentialLR can subclass this."""
    def __init__(
            self,
            schedulers: Sequence[LRScheduler],
            optimizer: Optional[Optimizer] = None,
    ):
        super().__init__(optimizer=optimizer or schedulers[0].optimizer)
        self.schedulers = schedulers

    def state_dict(self):
        return [scheduler.state_dict() for scheduler in self.schedulers]

    def load_state_dict(self, state_dicts):
        for scheduler, state_dict in zip(self.schedulers, state_dicts):
            scheduler.load_state_dict(state_dict)
```
